terraform/java/ec2/ubuntu/main.tf

```
@@ -0,0 +1,334 @@
+  }
+  filter {
+    name   = "state"
+    values = ["available"]
+  }
+  filter {
+    name   = "architecture"
+    values = [var.cpu_architecture]
+  }
+  filter {
+    name   = "image-type"
+    values = ["machine"]
+  }
+
+  filter {
+    name   = "root-device-name"
+    values = ["/dev/sda1"]
+  }
+
+  filter {
+    name   = "root-device-type"
+    values = ["ebs"]
+  }
+
+  filter {
+    name   = "virtualization-type"
+    values = ["hvm"]
+  }
+}
+
+resource "aws_instance" "main_service_instance" {
+  ami                                   = data.aws_ami.ami.id # Amazon Linux 2 (free tier)
+  instance_type                         = var.cpu_architecture == "x86_64" ? "t3.micro" : "t4g.micro"
+  key_name                              = local.ssh_key_name
+  iam_instance_profile                  = "APP_SIGNALS_EC2_TEST_ROLE"
+  vpc_security_group_ids                = [aws_default_vpc.default.default_security_group_id]
+  associate_public_ip_address           = true
+  instance_initiated_shutdown_behavior  = "terminate"
+
+  metadata_options {
+    http_tokens = "required"
+  }
+
+  root_block_device {
+    volume_size = 8
+  }
+
+  tags = {
+    Name = "main-service-${var.test_id}"
+  }
+}
+
+resource "null_resource" "main_service_setup" {
+  connection {
+    type = "ssh"
+    user = var.user
+    private_key = local.private_key_content
+    host = aws_instance.main_service_instance.public_ip
+  }
+
+  provisioner "remote-exec" {
+    inline = [
+      <<-EOF
+      #!/bin/bash
+      # Make the Terraform fail if any step throws an error
+      set -o errexit
+      # Install wget and Java based on OS
+      # Ubuntu commands
+      sudo apt-get update
+      sudo apt-get install -y software-properties-common
+      sudo add-apt-repository -y ppa:openjdk-r/ppa
+      sudo apt-get update
+      sudo apt-get install wget -y
+      # Try to fix broken packages if any
+      sudo apt-get install -y -f
+      # Clean apt cache
+      sudo apt-get clean
+      sudo apt-get autoremove -y
+      if [[ "${var.language_version}" == "8" ]]; then
+        sudo apt-get install -y openjdk-8-jdk 
+      else
+        sudo apt-get install -y openjdk-${var.language_version}-jdk
+      fi
+      sudo apt-get install -y awscli
+      
+      # Copy in CW Agent configuration
+      agent_config='${replace(replace(file("./amazon-cloudwatch-agent.json"), "/\\s+/", ""), "$REGION", var.aws_region)}'
+      echo $agent_config > amazon-cloudwatch-agent.json
+
+      # Get and run CW agent rpm/ubuntu
+      ${var.get_cw_agent_deb_command}
+      sudo dpkg -i -E ./cw-agent.deb
+      sudo /opt/aws/amazon-cloudwatch-agent/bin/amazon-cloudwatch-agent-ctl -a fetch-config -m ec2 -s -c file:./amazon-cloudwatch-agent.json
+
+      # Get ADOT
+      ${var.get_adot_jar_command}
+
+      # Get and run the sample application with configuration
+      aws s3 cp ${var.sample_app_jar} ./main-service.jar
+
+      JAVA_TOOL_OPTIONS=' -javaagent:/home/ubuntu/adot.jar' \
```

Are we certain the java agent will always be installed here? Can we just use relative path?
```suggestion
      JAVA_TOOL_OPTIONS=' -javaagent:./adot.jar' \
```

---

We have a similar path for ec2-user, that is why I used this

---

Regardless of previous code, let's see if we can make this improvement. If the default repository changes in the future this line will fail.